experiments/dsl/src/main/scala/org/apache/gearpump/streaming/dsl/plan/Planner.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.gearpump.streaming.dsl.plan
+
+import akka.actor.ActorSystem
+import org.apache.gearpump.partitioner.{CoLocationPartitioner, HashPartitioner, Partitioner}
+import org.apache.gearpump.streaming.TaskDescription
+import org.apache.gearpump.streaming.dsl.op._
+import org.apache.gearpump.streaming.dsl.partitioner.GroupByPartitioner
+import org.apache.gearpump.util.Graph
+
+import scala.collection.JavaConverters._
+
+class Planner {
```

Comment:
I like the overall approach where Graph[Op,OpEdge] is mapped to Graph[Task,Partitioner] and implemented by the Planner. It does make it difficult to combine APIs however Eg flatmap with ~, ~>, my original approach had been to treat TaskDescription as a monad which would allow the 2 DSLs to mix. I do think there are some places where type classes can be used and there may be a way still to combine DSLs. But there is no doubt it is an elegant design for some key functionality.